File: kattis/ProvingEquivalences.py
# link: https://open.kattis.com/problems/equivalences
# name: Proving Equivalences
__author__ = 'mislav'
import sys
visited = []
order = []

sys.setrecursionlimit(50000)


def dfs(node):
    global visited, order

    visited[node] = True
    if node in graph:
        for neighbour in graph[node]:
            if visited[neighbour]:
                continue
            dfs(neighbour)

    order.append(node)


def load_graph():

    input_line = raw_input()
    inputs = [int(s) for s in input_line.split()]

    vertices_count = inputs[0]
    edges_count = inputs[1]
    graph = {}

    for vertex in xrange(vertices_count):
        graph[vertex + 1] = []

    for _ in xrange(edges_count):
        input_line = raw_input()
        inputs = [int(s) for s in input_line.split()]
        neighbours = graph[inputs[0]]
        neighbours.append(inputs[1])
        graph[inputs[0]] = neighbours

    return graph


def reverse_graph(graph):

    graph_rev = {}
    for x in xrange(len(graph)):
        graph_rev[x+1] = []

    for first in graph.keys():
        for neighbour in graph[first]:
            current = graph_rev[neighbour]
            current.append(first)
            graph_rev[neighbour] = current

    return graph_rev

if __name__ == "__main__":

    test_cases = int(raw_input())
    for _ in xrange(test_cases):


        graph = load_graph()
        visited = [False] * (len(graph) + 1)
        order = []

        for node in graph.keys():
            if visited[node]:
                continue

            dfs(node)

        graph_rev = reverse_graph(graph)

        visited = [False] * (len(graph_rev) + 1)
        result = [0] * (len(graph_rev)+1)

        for index in xrange(len(order), 0, -1):

            start_node = order[index-1]
            if visited[start_node]:
                continue

            stack = [start_node]
            while len(stack) > 0:
                current = stack.pop()
                if visited[current]:
                    continue

                result[current] = start_node
                visited[current] = True

                for neighbour in graph_rev[current]:
                    stack.append(neighbour)

        dag = {}

        sinks = set()
        sources = set()

        for x in xrange(len(graph)):

            vertex = result[x+1]
            sources.add(vertex)
            sinks.add(vertex)
            neighbours = dag.get(vertex, set())
            for neighbour in graph[vertex]:
                if result[neighbour] != vertex:
                    neighbours.add(neighbour)

            dag[vertex] = neighbours

        if len(dag) == 1:
            print 0
            continue

        for key, value in dag.iteritems():
            if len(value) > 0:
                sinks.discard(key)
                for source in value:
                    sources.discard(source)

        print max(len(sinks), len(sources))

File: kattis/ProvingEquivalences.java
// link: https://open.kattis.com/problems/equivalences
// name: Proving Equivalences
package kattis;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.Scanner;

public class ProvingEquivalences {

	private static Scanner input;

	private static int cnt, scnt;
	private static int[] id, pre, low;
	private static Deque<Integer> stack;
	private static ArrayList<ArrayList<Integer>> graph;

	public static void main(String[] args) {
		input = new Scanner(System.in);

		int testsLeft = input.nextInt();
		while (testsLeft-- > 0) {

			solve();
		}
		input.close();
	}

	private static void solve() {
		cnt = 0;
		scnt = 0;
		int vertexCount = input.nextInt();
		int edgeCount = input.nextInt();

		graph = new ArrayList<>();

		for (int i = 0; i < vertexCount + 1; ++i) {
			ArrayList<Integer> neighbours = new ArrayList<>();
			graph.add(neighbours);
		}

		for (int i = 0; i < edgeCount; ++i) {
			int from = input.nextInt();
			int to = input.nextInt();
			graph.get(from).add(to);
		}

		stack = new ArrayDeque<>();
		id = new int[vertexCount + 1];
		pre = new int[vertexCount + 1];
		low = new int[vertexCount + 1];

		for (int i = 0; i < vertexCount + 1; ++i) {
			id[i] = -1;
			pre[i] = -1;
			low[i] = -1;
		}

		for (int v = vertexCount; v > 0; --v) {
			if (pre[v] == -1)
				strong_components(v);
		}
		
		int[] outDegrees = new int[scnt];
		int[] inDegrees = new int[scnt];

		for (int from = 0; from < vertexCount + 1; ++from) {
			for (Integer to : graph.get(from)) {

				if (id[from] == id[to]) {
					continue;
				}

				++outDegrees[id[from]];
				++inDegrees[id[to]];
			}
		}
		
		if (scnt == 1){
			System.out.println(0);
			return;
		}
		
		int sinks = 0;
		int sources = 0;

		for (int v = scnt - 1; v >= 0; --v) {
			if (outDegrees[v] == 0){
				++sinks;
			}
			
			if (inDegrees[v] == 0){
				++sources;
			}
		}
		int result = sinks > sources ? sinks : sources;

		System.out.println(result);
	}

	private static void strong_components(int w) {

		int t;
		int min = cnt;
		low[w] = cnt;
		pre[w] = cnt;
		++cnt;

		stack.push(w);

		ArrayList<Integer> neighbours = graph.get(w);

		for (Integer neighbour : neighbours) {

			if (pre[neighbour] == -1) {
				strong_components(neighbour);
			}

			if (low[neighbour] < min) {
				min = low[neighbour];
			}
		}

		if (min < low[w]) {
			low[w] = min;
			return;
		}

		while (true) {

			t = stack.pop();

			id[t] = scnt;
			low[t] = graph.size();
			if (t == w) {
				break;
			}
		}
		scnt++;
	}
}
